src/ir/validate.rs

```diff
@@ -0,0 +1,451 @@
+use crate::{ir, ir::Ctx};
+
+use super::IndexStore;
+/// Implements validation checks for IR data structures.
+/// If calling the methods in this does not result in a panic, then the corresponding IR structure is valid.
+/// The validity condition for each structure is defined in the corresponding method.
+pub struct Validate<'a> {
+    /// The component being validated.
+    comp: &'a ir::Component,
+    /// The context for the program being evaluated
+    ctx: &'a IndexStore<ir::Component>,
+}
+
+impl<'a> Validate<'a> {
+    /// Validate the entire component
+    pub fn comp(&self) {
+        // Validate exprs
+        for (eidx, _) in self.comp.exprs().iter() {
+            self.expr(eidx);
+        }
+
+        // Validate times
+        for (tidx, _) in self.comp.times().iter() {
+            self.time(tidx);
+        }
+
+        // Validate props
+        for (pidx, _) in self.comp.props().iter() {
+            self.prop(pidx);
+        }
+
+        // Validate ports
+        for (pidx, _) in self.comp.ports().iter() {
+            self.port(pidx);
+        }
+
+        // Validate params
+        for (pidx, _) in self.comp.params().iter() {
+            self.param(pidx);
+        }
+
+        // Validate events
+        for (evidx, _) in self.comp.events().iter() {
+            self.event(evidx);
+        }
+
+        // Validate invokes
+        for (iidx, _) in self.comp.invocations().iter() {
+            self.invoke(iidx);
+        }
+
+        // Validate instances
+        for (iidx, _) in self.comp.instances().iter() {
+            self.instance(iidx);
+        }
+
+        // Validate commands
+        for cmd in self.comp.cmds.iter() {
+            self.command(cmd.clone());
+        }
+    }
+
+    /// An Expr is valid if:
+    /// (1) It is bound in the component
+    /// (2) All the exprs it uses are also bound
+    pub fn expr(&self, eidx: ir::ExprIdx) {
```

Comment:
This can get a bit expense but here is an invariant in the construction of expressions that we can potentially use: if `%n` is defined, then all expressions `%i` where `i` < `n` are also defined. Furthermore, a compound expression's index (for example `ir::Expr::Bin`) has a higher index than its children because the children must be constructed before the compound expression. This means, we should only have to check if the expression is defined in the and ignore its children if the underlying expression is a compound.

Reply:
Yeah, kind of tricky but the well-formedness constraint for `IndexStore` is that it doesn't even give you `%2` till you've constructed `%1` and `%0`. This is a constraint that cloning an `Interned` should maintain is more of a property of the intern store than the component.

Think of it this way: the validation pass is responsible for checking relations *between* different tables while each table is responsible for maintaining its own variants. Does that sound somewhat reasonable to you?